### PR TITLE
Move to use a function for startup instead of a class

### PR DESCRIPTION
**Pull request by TimOrme**
This allows easier logical construction of worker settings, and supports a stricter type structure for the settings themselves.

In doing this, brought up some other questions, mostly around logging.  Was thinking this might be a nice way to let users specify their own logging configuration, but it comes with a couple problems there:

1. We can't really provide a sensible default logging config for them.
2. It becomes a bit less clear where worker "startup" code should go.  There is already a hook for startup, but it happens after the worker loop is running, and therefore isn't a great place to inject logging, as other logs have already occurred.

I also ran into some issues that are still standing in this PR.

1.  Tests fail for the CLI.  Even though I am configuring logging, and the log statements are hit, it appears that no output is captured for some reason.
2. There is a type error with the WorkerCoroutine protocol.  I can't figure out a good way to work around this, but for whatever reason, prototyping a `__call__` method doesn't seem to work as expected with mypy.

Other eyes on the above would be helpful.

Once we those sorted out, I can update the documentation to reflect the changes as well.

### CODE REVIEW

File: narq/worker.py [hunk 1]
```diff
@@ -45,6 +45,34 @@ class Function:
     max_tries: Optional[int]
 
 
+@dataclass
+class WorkerSettings:
```
Comment (inactivist) [inline narq/worker.py:49]: LGTM

File: narq/cli.py [hunk 2]
```diff
@@ -24,19 +18,23 @@
 @click.argument('worker-settings', type=str, required=True)
```
Comment (inactivist) [inline narq/cli.py:18]: Feels like we could have a better name for this arg, based on what I see.  It's the name of a callable that:

1. Performs additional startup magic as desired (logging config, in your example and test code), and
2. returns a new instance of `WorkerSettings` dataclass to be used elsewhere

so `worker-settings` seems a bit misleading (though I suppose we can document around it).

File: tests/test_cli.py [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
 
 
 async def foobar(ctx):
     return 42
 
 
-class WorkerSettings:
-    burst = True
-    functions = [foobar]
+def worker_start():
```
Comment (inactivist) [inline tests/test_cli.py:13]: See, this example seems odd to me: it doesn't start the worker at all.

It performs unrelated system initialization, and returns `WorkerSettings` to be used elsewhere.

I know this is a unit test, but I think we can help document our intentions here and elsewhere.

**Review by inactivist — COMMENTED** [review]
I think the basic goal is a good one, with a few minor concerns about naming and communicating intent via documentation.

**Comment by inactivist** [comment]
> In doing this, brought up some other questions, mostly around logging. Was thinking this might be a nice way to let users specify their own logging configuration, but it comes with a couple problems there:
> 
> 1. We can't really provide a sensible default logging config for them.

It's fairly common to init logging from a dict or a config file; Flask lets you do whatever you want before starting the app;[Django uses the latter method](https://docs.djangoproject.com/en/3.1/topics/logging/#configuring-logging), in the project settings.py file.

Maybe easier in this project,with a specific callback to init logging; otherwise, use the default settings?

**Comment by inactivist** [comment]
> 2\. It becomes a bit less clear where worker "startup" code should go.  There is already a hook for startup, but it happens after the worker loop is running, and therefore isn't a great place to inject logging, as other logs have already occurred.

so maybe there should b pre-startup and post-startup

**Comment by inactivist** [comment]
Oops, hit the wrong button while writing a comment; re-opening the PR

**Comment by TimOrme** [comment]
> > In doing this, brought up some other questions, mostly around logging. Was thinking this might be a nice way to let users specify their own logging configuration, but it comes with a couple problems there:
> > 
> > 1. We can't really provide a sensible default logging config for them.
> 
> It's fairly common to init logging from a dict or a config file; Flask lets you do whatever you want before starting the app; [Django uses the latter method](https://docs.djangoproject.com/en/3.1/topics/logging/#configuring-logging), in the project settings.py file.
> 
> Maybe easier in this project to provide a specific callback to init logging; otherwise, use the worker's default settings?

Yeah I had considered this as well.  Would it make sense to have two configuration options here then?  One to retrieve the worker settings, and then one to optionally provide an override for default logging settings?  It makes things very clear and specific about what there doing, though maybe too granular.  Perhaps it makes more sense to just have two methods:

* one to retrieve worker settings
* one top optionally override a "pre-start" hook.  We could provide a default which only configures logging.

I think it feels a bit over-specific to label the second one as only for logging since we dont really have a way to restrict what that method does anyhow.

File: narq/cli.py (hunk 2 above)
Comment (TimOrme) [reply narq/cli.py:18]: Agreed on all fronts.  See comment below which I think at least helps address the multiple concerns aspect, but maybe it does make sense to start shelling out some documentation along with this, just to see how that looks with it.  Sometimes easier to evaluate how good the decision is based on how easy it is to explain.

**Comment by inactivist** [comment]
> * one to retrieve worker settings
> * one top optionally override a "pre-start" hook.  We could provide a default which only configures logging.

I like the sound of that, though of course I'd like to see a code snippet or two showing how it'd be used.

**Comment by inactivist** [comment]
LGTM, do we have an example overriding logging settings in the new scheme?

**Review by inactivist — APPROVED** [review]

**Comment by inactivist** [comment]
LGTM, do we have an example overriding logging settings in the new scheme?

**Comment by TimOrme** [comment]
> LGTM, do we have an example overriding logging settings in the new scheme?

Think you found it but added one here: https://github.com/codicide/narq/pull/26/files#diff-c9072fc5e400185aff19afa61b1d0a3dca00dd015588f2ec360dc9ef7cc75998R29-R35

Also added some text-docs with a more explicit example of it here: https://github.com/codicide/narq/pull/26/files#diff-8d068e8797e88947c320f79e856c3e16a72b730124a8f9d7031e2c4680dfa534R108-R119

**Pull request merged**